src/main/java/com/google/devtools/build/lib/skyframe/IgnoredPackagePrefixesValue.java

```diff
@@ -54,19 +71,33 @@ public static SkyKey key(RepositoryName repository) {
   }
 
   public ImmutableSet<PathFragment> getPatterns() {
-    return patterns;
+    return ignoredPrefixes;
```

Comment:
Thanks for the suggestion! Any advice on changing the place that doesn't perform the check immediately but pass the set of ignored directory PathFragment to others? For example, in `TargetPatternFunction`, it passes the set to `TargetPattern`. Then `TargetPattern` use the set to produce an `IgnoredPathFragmentsInScopeOrFilteringIgnorer` (through `getAllIgnoredSubdirectoriesToExclude`) and pass to `TargetPatternResolver`, etc.

Reply:
I would pass a predicate instead of the set. The predicate can be the method reference ::isPathFragmentIgnored. Is there any place that actually needs to iterate over the set, versus checking for membership?

Reply:
I think there are some places. Examples:
Line 174-17 in EnvironmentBackedRecursivePackageProvider.java & Line 46-4 in RecursivePkgValueRootPackageExtractor.java
```
ImmutableSet<PathFragment> filteredIgnoredSubdirectories =
        ImmutableSet.copyOf(
            Iterables.filter(ignoredSubdirectories, path -> path.startsWith(directory)));
```

Reply:
Any suggestion @janakd? Thanks!

Reply:
Sorry for the late reply. It looks like this feature would require a fair amount of work to the rest of Bazel to actually work as intended, but I'm open to it. In particular, the fancy "filtering" code you pointed to would no longer work trivially. I think you could make it work by enriching the Predicate you're providing to have a method #restrictedTo(PathFragment fragment). That method would filter the prefixes, and leave the regexes unchanged, since it would be fairly hard to test them.

Reply:
It looks like you can do the filtering of the ignored paths as before, and just propagate any regexes along unchanged (since it would be very difficult to see if a regex _could_ apply to a subdirectory). You'd probably want to have special cases for the "totally ignored," "empty ignored paths and empty regexes," and "non-empty ignored paths and empty regexes." You might want to reverse the logic, so the "filterer" takes in a target pattern and has to produce a new filterer, rather than putting the target pattern in charge, but I could see either way.